test/e2e/autoscaling/custom_metrics_stackdriver_autoscaling.go

```diff
@@ -254,6 +374,77 @@ func objectHPA(namespace string, metricTarget int64) *as.HorizontalPodAutoscaler
 	}
 }
 
+type externalMetricTarget struct {
+	value     int64
+	isAverage bool
+}
+
+func externalHPA(namespace string, metricTargets map[string]externalMetricTarget) *as.HorizontalPodAutoscaler {
+	var minReplicas int32 = 1
+	metricSpecs := []as.MetricSpec{}
+	selector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{"resource.type": "gke_container"},
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      "resource.labels.namespace_id",
+				Operator: metav1.LabelSelectorOpIn,
+				// TODO(bskiba): change default to real namespace name it is available
+				// from Stackdriver.
+				Values: []string{"default", "dummy"},
+			},
+			{
+				Key:      "resource.labels.pod_id",
+				Operator: metav1.LabelSelectorOpExists,
+				Values:   []string{},
+			},
+			{
+				Key:      "resource.labels.zone",
+				Operator: metav1.LabelSelectorOpNotIn,
+				Values:   []string{"dummy", "dummy2"},
+			},
+			{
+				Key:      "resource.labels.cluster_name",
+				Operator: metav1.LabelSelectorOpNotIn,
+				Values:   []string{"foo", "bar"},
+			},
+		},
+	}
+	for metric, target := range metricTargets {
+		var metricSpec as.MetricSpec
+		metricSpec = as.MetricSpec{
+			Type: as.ExternalMetricSourceType,
+			External: &as.ExternalMetricSource{
+				MetricName:     "custom.googleapis.com|" + metric,
+				MetricSelector: selector,
+			},
+		}
+		if target.isAverage {
+			metricSpec.External.TargetAverageValue = resource.NewQuantity(target.value, resource.DecimalSI)
+		} else {
+			metricSpec.External.TargetValue = resource.NewQuantity(target.value, resource.DecimalSI)
+		}
+		metricSpecs = append(metricSpecs, metricSpec)
+	}
+	hpa := &as.HorizontalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-metrics-external-hpa",
```

Comment:
nit: I know external metrics are also custom, and we use the same adapter for both, but as we're using "custom metrics" and "external metrics" to refer to separate APIs, can we remove "custom" part of the name?

Reply:
We can, but I think it's a mistake. I don't like the fact that we draw the line between the metrics looking at the APIs instead of what the metrics mean from user's point of view. But probably not everyone agrees and this is not the right place to decide this. If you're feeling strongly about this I'll remove it:)

Reply:
Not that strongly:)